--- a/configure.ac
+++ b/configure.ac
@@ -992,6 +992,16 @@ case $host in
     cc_cflags="-O2 -o32"   # no -g, it disables all optimizations
     # this suits both mips32 and mips64
     GMP_INCLUDE_MPN(mips32/mips-defs.m4)
+    
+    mips_isa_rev=$($CC -dM -E - < /dev/null | grep "__mips_isa_rev"| cut -d ' '  -f3)
+
+    if test -z "$mips_isa_rev"; then
+      mips_isa_rev=0
+    fi
+
+    if test $mips_isa_rev -ge 6; then
+      path="mips32/r6"
+    fi
 
     case $host in
       [mips64*-*-* | mips*-*-irix[6789]*])
@@ -1013,7 +1023,13 @@ case $host in
         cc_64_cflags="-O2 -64"		# no -g, it disables all optimizations
         cc_64_ldflags="-Wc,-64"
         path_64="mips64"
-        ;;
+        
+	if test $mips_isa_rev -ge 6; then
+          path_n32="mips64/r6"
+          path_64="mips64/r6"
+        fi
+    	
+	;;
     esac
     ;;
 
